### ISSUE TITLE
[Bug]: Issues with Addons: kube-state-metrics, aws-cloudwatch-metrics, node-termination-handler

### ISSUE THREAD

**Issue by armenr** [aws-ia/terraform-aws-eks-blueprints#464]
### Welcome to Amazon EKS Blueprints!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/aws-ia/terraform-aws-eks-blueprints/issues) and didn't find any.

### Amazon EKS Blueprints Release version

v4.0.2

### What is your environment, configuration and the example used?

Terraform v1.1.9 on darwin_arm64

### My Addons Module

```

module "eks_blueprints_kubernetes_addons" {
  source = "github.com/aws-ia/terraform-aws-eks-blueprints//modules/kubernetes-addons?ref=v4.0.2"

  depends_on = [
    module.eks-blueprints.managed_node_groups,
    data.aws_eks_cluster.is_ready
  ]

  eks_cluster_id = module.eks-blueprints.eks_cluster_id

  # EKS Addons
  ## EBS-CSI Driver
  enable_amazon_eks_aws_ebs_csi_driver    = true
  amazon_eks_aws_ebs_csi_driver_config    = {
    addon_name               = "aws-ebs-csi-driver"
    addon_version            = "v1.5.3-eksbuild.1"
    service_account          = "ebs-csi-controller-sa"
    resolve_conflicts        = "OVERWRITE"
    namespace                = "kube-system"
    # additional_iam_policies  = []
    # service_account_role_arn = ""
    # tags                     = {}
  }

  ## CoreDNS
  enable_amazon_eks_coredns               = true
  amazon_eks_coredns_config               = {
    addon_name               = "coredns"
    addon_version            = "v1.8.7-eksbuild.1"
    service_account          = "kube-proxy"
    resolve_conflicts        = "OVERWRITE"
    namespace                = "kube-system"
    # additional_iam_policies  = []
    # service_account_role_arn = ""
    # tags                     = {}
  }

  ## Kube-Proxy
  enable_amazon_eks_kube_proxy            = true
  amazon_eks_kube_proxy_config            = {
    addon_name               = "kube-proxy"
    addon_version            = "v1.22.6-eksbuild.1"
    service_account          = "kube-proxy"
    resolve_conflicts        = "OVERWRITE"
    namespace                = "kube-system"
    # additional_iam_policies  = []
    # service_account_role_arn = ""
    # tags                     = {}
  }

  ## VPC-CNI
  enable_amazon_eks_vpc_cni               = true
  amazon_eks_vpc_cni_config               = {
    addon_name               = "vpc-cni"
    addon_version            = "v1.11.0-eksbuild.1"
    service_account          = "aws-node"
    resolve_conflicts        = "OVERWRITE"
    namespace                = "kube-system"
    # additional_iam_policies  = []
    # service_account_role_arn = ""
    # tags                     = {}
  }

  # Why not?
  enable_metrics_server                   = true
  # enable_aws_cloudwatch_metrics           = true
  # enable_kube_state_metrics               = true
  # enable_aws_node_termination_handler     = true

  enable_amazon_prometheus                = true
  amazon_prometheus_workspace_endpoint    = module.eks_blueprints_kubernetes_addons.amazon_prometheus_workspace_endpoint

  enable_prometheus = true

  # enable_aws_load_balancer_controller  = true
  # enable_external_dns                  = true
  #K8s Add-ons
  # enable_argocd                       = true
  # enable_aws_for_fluentbit            = true
  # enable_cluster_autoscaler           = true
  # enable_prometheus                   = true
}

```

### What did you do and What did you see instead?

I ran a terraform plan/apply, and a few unexpected things happen:

1. `enable_aws_node_termination_handler     = true` - this produces *no* effect. That is to say, I don't even see Terraform detect/attempt to provision any resources for this
2. For `kube-state-metrics` & `aws-cloudwatch-metrics` --> this is the output I get, which tells me that maybe these vars are not defined/included in the vars for the`eks-blueprints` / `kubernetes-addons` submodule

```
│ Error: Unsupported argument
│
│   on addons-kube-system.tf line 67, in module "eks_blueprints_kubernetes_addons":
│   67:   enable_aws_cloudwatch_metrics           = true
│
│ An argument named "enable_aws_cloudwatch_metrics" is not expected here.
╵
╷
│ Error: Unsupported argument
│
│   on addons-kube-system.tf line 68, in module "eks_blueprints_kubernetes_addons":
│   68:   enable_kube_state_metrics               = true
│
│ An argument named "enable_kube_state_metrics" is not expected here.
╵
ERRO[0035] 1 error occurred:
	* exit status 1
```

### Additional Information

```shell
I'm using the blueprints module for also provisioning all of EKS, etc...and doing it the vanilla way.

I'm also successfully set up with Karpenter, using a `managed` nodegroup + AL2 x64
```

**Comment by vara-bonthu** [comment]
Thanks for reporting the issue @armenr 

1> `enable_aws_node_termination_handler = true` is not showing any effect due to the condition below. It's expecting additional input from `auto_scaling_group_names` from self-managed node groups. I think we can handle this better.

@askulkarni2 I think we can change the below condition to `||` so that users can just use aws_node_termination_handler if they want to and manage autoscaling lifecycle hooks outside our module.

https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/f22c583290b2321a2bb169f11b57e7bc9d83ccfa/modules/kubernetes-addons/main.tf#L99

2> `kube-state-metrics` currently disabled. We will fix the docs or add the `kube-state-metrics` add-on within this module

3> `aws-cloudwatch-metrics` added only after tag v.4.0.2. This should be available if you try with latest version that i published today v4.0.3

We will close this once we have the fixes for the above two issues

**Comment by armenr** [comment]
@vara-bonthu - Thank you for the update! I'm not sure if I'm introducing scope creep, but I did discover an additional issue with Karpenter.

https://github.com/aws-ia/terraform-aws-eks-blueprints/blob/main/modules/kubernetes-addons/karpenter/data.tf

^^ This needs to be updated to reflect the permissions included in the Karpenter Getting Started (https://karpenter.sh/v0.8.2/getting-started/getting-started-with-eksctl/cloudformation.yaml). 

Please also see: https://github.com/aws/karpenter/issues/1706

I am getting the same behavior (Karpenter stuck, does not have permission to delete launch templates).

Additionally, do you think it would be worthwhile to maybe allow the module to accept an "override" or "additional" permissions/policies set...that way we can allow the operator/builder to rely on defaults, but also permit flexibility/allow for operators to work through these issues without needing to depend on centralized module updates?

**Comment by vara-bonthu** [comment]
@armenr Happy for you to raise a separate issue for Karpenter that you described above. Provide the example that you used etc. We will take a look and provide feedback accordingly

**Comment by bryantbiggs** [comment]
The karpenter permissions should be resolved now in #546 - please let us know if this issue is now resolved with the current version of Blueprints @armenr

**Comment by bryantbiggs** [comment]
closing for now - please let us know if you have any further issues and we can re-open
